### DISCUSSION TITLE
Render to FBO

### DISCUSSION THREAD

**Discussion by jwinarske** [google/filament#4427] [Q&A]
I would like to build a GLTFW viewer that renders to a FBO.  OpenGL/GLES3 and shared context are already setup prior to initializing engine.  Can someone suggest a sample and general changes required to do this?  I'd be happy to upstream any useful part of it, even a working sample.

**Comment by jwinarske** [comment]
This is for use with Flutter as a Texture.   I'm thinking a new backend is needed, with a few functions exposed to set context.  Basically no platform code is needed as it would be handled external.

**Reply by romainguy** [reply]
You don’t need a new backend. On Android you can use a SurfaceTexture and have Filament render into the SurfaceTexture’s Surface. On iOS @bejado can explain what the equivalent would be.

**Comment by jwinarske** [comment]
This is for GLES3 on Linux/Wayland (no X11).

I've refactored FilamentApp removing SDL and Imgui deps. The new backend is to remove the SDL deps.  The flutter app becomes what imgui was.  So control points need exposing.

In the case of a Flutter texture plugin everything is configured before filament gets initialized, including shared context.

The new backend is to remove window interaction.  Basically all I need is the Flutter app:run in my render loop.  No native window anything, as that's handled in the flutter embedder.

Then there are cmake changes to remove X11 deps. This is on a system without X11 (Wayland).

FilamentApp could become a base class (no SDL, no Imgui), then be extended for SDL/Imgui?

**Reply by romainguy** [reply]
Just to be clear: filamentapp is a rather terrible implementation that grew organically for our samples and is in no way intended to be used for general purpose (it doesn't do vsync, etc.). Anyway, [Snapchat recently mentioned possibly upstreaming](https://github.com/google/filament/discussions/4410) their work for EGL on Linux, which would probably be helpful in your case.

For the use case you describe we have the concept of importing a texture ID, which is how you'd create a `RenderTarget` for that texture coming from Flutter.

**Reply by jwinarske** [reply]
EGL has already been initialized, and configured with a valid context before Filament enters the picture.  In addition I have an external process resolver.

So the plugin I'm creating has an init, de-init, render, and as-needed functions to control the model's front end.  The plugin doesn't need to do any EGL/platform setup/config, just render to a FBO.  Render is called in the main loop.

The goal was to move the Platform and EGL setup outside of the plugin, easier for third-parties, minimize load time, and eliminate redundant code.  I have a Rust app (glutin based) that loads and interacts with the plugin for development and visual testing.  The flutter embedder loads it at runtime as needed.

Quick and easy currently looks like a FilamentApp and EGL backend refactor.  Only I would prefer to push upstream as I believe it will benefit the Flutter community.  Our custom Flutter embedder **will be** open sourced in the near future.

**Reply by romainguy** [reply]
Why can’t you let Filament create its own context and use shared contexts? You shouldn’t need to change Filament’s EGL backend.

**Reply by jwinarske** [reply]
All contexts in use are shared, as that's what is required for the flutter engine.  My primary concern would be to minimize load time, and size (redundant code).

Is there an example that follows your suggestion?

**Reply by romainguy** [reply]
What redundant code are you talking about exactly? Filament expects its context to be initialized in very specific ways.

**Reply by jwinarske** [reply]
In my case platform/EGL setup code

**Reply by romainguy** [reply]
I'd need to see more of your code to understand what you call platform/EGL setup code. If you are talking about Filament's platform/EGL code, it's not redundant because like I said we need to control it.

**Reply by jwinarske** [reply]
I'll try using the shared context approach.  I just found some examples.  Is pbuffer support a hard requirement?
